packages/url_launcher/url_launcher_android/android/src/test/java/io/flutter/plugins/urllauncher/WebViewActivityTest.java

```diff
@@ -5,13 +5,124 @@
+  }
+
   @Test
   public void extractHeaders_returnsEmptyMapWhenHeadersBundleNull() {
     assertEquals(WebViewActivity.extractHeaders(null), Collections.emptyMap());
   }
+
+  @Test
+  public void httpRequest_doesNotStartActivity() {
+    boolean result = FlutterWebViewClient.urlShouldRunActivity(webView, "http://www.flutter.dev/");
+
+    verify(context, times(0)).startActivity(any(Intent.class));
+    assertFalse(result);
+  }
+
+  @Test
+  public void httpsRequest_doesNotStartActivity() {
+    boolean result = FlutterWebViewClient.urlShouldRunActivity(webView, "https://www.flutter.dev/");
+
+    verify(context, times(0)).startActivity(any(Intent.class));
+    assertFalse(result);
+  }
+
+  @Test
+  public void smsLink_startsActivity() {
+    boolean result = FlutterWebViewClient.urlShouldRunActivity(webView, "sms:1234567890");
+
+    verify(context, times(1)).startActivity(any(Intent.class));
+    assertTrue(result);
+  }
+
+  @Test
+  public void telLink_startsActivity() {
+    boolean result = FlutterWebViewClient.urlShouldRunActivity(webView, "tel:1234567890");
+
+    verify(context, times(1)).startActivity(any(Intent.class));
+    assertTrue(result);
+  }
+
+  @Test
+  public void geoLink_startsActivity() {
+    boolean result = FlutterWebViewClient.urlShouldRunActivity(webView, "geo:34.090335,-84.255769");
+
+    verify(context, times(1)).startActivity(any(Intent.class));
+    assertTrue(result);
+  }
+
+  @Test
+  public void marketLink_startsActivity() {
+    boolean result =
+        FlutterWebViewClient.urlShouldRunActivity(webView, "market://details?id=com.example.test");
+
+    verify(context, times(1)).startActivity(any(Intent.class));
+    assertTrue(result);
+  }
+
+  @Test
+  public void contentLink_startsActivity() {
+    boolean result =
+        FlutterWebViewClient.urlShouldRunActivity(
+            webView, "content://media/external/images/media/");
+
+    verify(context, times(1)).startActivity(any(Intent.class));
+    assertTrue(result);
+  }
+
+  @Test
+  public void fileLink_startsActivity() {
+    boolean result =
+        FlutterWebViewClient.urlShouldRunActivity(webView, "file:///sdcard/download/example.jpg");
+
+    verify(context, times(1)).startActivity(any(Intent.class));
+    assertTrue(result);
+  }
+
+  @Test
+  public void onLoadResource_allowsWebPageLoad() {
+    String url = "https://www.flutter.dev/";
+    webView.getWebViewClient().onLoadResource(webView, url);
+
+    verify(context, times(0)).startActivity(any(Intent.class));
+    assertFalse(FlutterWebViewClient.resourceShouldOpenDocument(webView, url));
+  }
+
+  @Test
+  public void onLoadResource_loadsPdf() {
+    String url = "https://www.flutter.dev/test.pdf";
+    webView.getWebViewClient().onLoadResource(webView, url);
+
+    verify(context, times(1)).startActivity(any(Intent.class));
+    assertTrue(FlutterWebViewClient.resourceShouldOpenDocument(webView, url));
+  }
```

Comment:
Unit tests to make sure that the URL launcher parses the URL properly based on scenario (.pdf, :mailto, etc)